gateway/ln-gateway/src/lib.rs

```diff
@@ -1590,61 +1591,58 @@ impl Gateway {
     async fn send_payment_v2(
         &self,
         payload: SendPaymentPayload,
-    ) -> std::result::Result<std::result::Result<[u8; 32], Signature>, SendPaymentErrorV2> {
+    ) -> anyhow::Result<std::result::Result<[u8; 32], Signature>> {
         let clients = self.clients.read().await;
 
         let client = clients
             .get(&payload.federation_id)
-            .ok_or(SendPaymentErrorV2::UnknownFederationId)?
+            .ok_or(anyhow!("Federation client not available"))?
             .value();
 
-        let operation_id = OperationId(payload.contract.contract_id().0.into_inner());
+        // The operation id is equal to the contract id which also doubles as the
+        // message signed by the gateway via the forfeit signature to forfeit
+        // the gateways claim to a contract in case of cancellation. We only create a
+        // forfeit signature after the we have started the send state machine to
+        // prevent replay attacks with a previously cancelled outgoing contract
+        let operation_id = OperationId::from_encodable(payload.contract.clone());
 
         let module = client.get_first_module::<GatewayClientModuleV2>();
 
         if client.operation_exists(operation_id).await {
             return Ok(module.subscribe_send(operation_id, payload.contract).await);
         }
 
+        // Since the following three checks may only fail due to client side
+        // programming error we do not have to enable cancellation and can check
+        // them before we start the state machine.
         if payload.contract.claim_pk != module.keypair.public_key() {
-            return Err(SendPaymentErrorV2::NotOurKey);
+            bail!("The outgoing contract keyed to another gateway");
         }
 
         if payload.invoice.consensus_hash::<sha256::Hash>() != payload.contract.invoice_hash {
-            return Err(SendPaymentErrorV2::InvalidInvoiceHash);
+            bail!("The invoices hash does not match the contract");
         }
 
+        let invoice_msats = payload
+            .invoice
+            .amount_milli_satoshis()
+            .ok_or(anyhow!("Invoice is missing amount"))?;
```

Comment:
Here you make the assumption that the client will check that before and can display an error based on that preliminary check, which is valid imo.

Reply:
Yes, the client checks that and has a structured Error variant for it.